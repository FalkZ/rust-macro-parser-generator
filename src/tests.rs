macro_rules! tests {
    (@debug: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_debug_snapshot!(&r.$t);
        }
    };
    (@display: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_display_snapshot!(&r.$t);
        }
    };
    // `()` indicates that the macro takes no argument.
    ($($name:ident: $str:literal),*) => {
        mod tokens {
            $(
                tests!(@debug: $name, tokens, $str);
            )*
        }
        mod statements {
            $(
                tests!(@debug: $name, statements, $str);
            )*
        }
        mod typescript {
            $(
                tests!(@display: $name, typescript, $str);
            )*
        }
    };
}

tests!(
    enums: "./examples/m1n/Enum.m1n",
    class: "./examples/m1n/Class.m1n",
    file: "./examples/m1n/file.m1n",
    math: "./examples/m1n/std/math.m1n",
    matcher: "./examples/m1n/match.m1n",
    assignment: "./examples/m1n/assignment.m1n",
    operator: "./examples/m1n/operator.m1n"
);
