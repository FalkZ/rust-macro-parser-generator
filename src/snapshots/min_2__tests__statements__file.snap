---
source: src/tests.rs
expression: "&r.statements"
---
[
    statements {
        position: Position {
            column: 0,
            line: 1,
        },
        statement: imports(
            imports {
                position: Position {
                    column: 0,
                    line: 1,
                },
                imports: [
                    import {
                        position: Position {
                            column: 4,
                            line: 2,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/util`",
                                position: Position {
                                    column: 4,
                                    line: 2,
                                },
                            },
                        ),
                        import_items: Some(
                            import_items {
                                position: Position {
                                    column: 15,
                                    line: 2,
                                },
                                items: [
                                    import_item {
                                        position: Position {
                                            column: 17,
                                            line: 2,
                                        },
                                        item: IDENT(
                                            Token {
                                                raw: "pipe",
                                                position: Position {
                                                    column: 17,
                                                    line: 2,
                                                },
                                            },
                                        ),
                                    },
                                    import_item {
                                        position: Position {
                                            column: 23,
                                            line: 2,
                                        },
                                        item: IDENT(
                                            Token {
                                                raw: "match",
                                                position: Position {
                                                    column: 23,
                                                    line: 2,
                                                },
                                            },
                                        ),
                                    },
                                ],
                            },
                        ),
                    },
                    import {
                        position: Position {
                            column: 4,
                            line: 3,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/math`",
                                position: Position {
                                    column: 4,
                                    line: 3,
                                },
                            },
                        ),
                        import_items: None,
                    },
                ],
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 4,
        },
        statement: definition(
            definition {
                position: Position {
                    column: 0,
                    line: 4,
                },
                name: IDENT(
                    Token {
                        raw: "version",
                        position: Position {
                            column: 0,
                            line: 4,
                        },
                    },
                ),
                arguments: [
                    calls {
                        position: Position {
                            column: 9,
                            line: 4,
                        },
                        argument: body {
                            position: Position {
                                column: 9,
                                line: 4,
                            },
                            value: primitive_value(
                                NUMBER(
                                    Token {
                                        raw: "0",
                                        position: Position {
                                            column: 9,
                                            line: 4,
                                        },
                                    },
                                ),
                            ),
                            expressions: [],
                        },
                    },
                ],
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 5,
        },
        statement: definition(
            definition {
                position: Position {
                    column: 0,
                    line: 5,
                },
                name: IDENT(
                    Token {
                        raw: "derive",
                        position: Position {
                            column: 0,
                            line: 5,
                        },
                    },
                ),
                arguments: [
                    calls {
                        position: Position {
                            column: 8,
                            line: 5,
                        },
                        argument: body {
                            position: Position {
                                column: 8,
                                line: 5,
                            },
                            value: primitive_value(
                                IDENT(
                                    Token {
                                        raw: "equal",
                                        position: Position {
                                            column: 8,
                                            line: 5,
                                        },
                                    },
                                ),
                            ),
                            expressions: [],
                        },
                    },
                ],
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 7,
        },
        statement: variable(
            variable {
                position: Position {
                    column: 0,
                    line: 7,
                },
                modifiers: [],
                name: IDENT(
                    Token {
                        raw: "PI",
                        position: Position {
                            column: 0,
                            line: 7,
                        },
                    },
                ),
                body: body {
                    position: Position {
                        column: 5,
                        line: 7,
                    },
                    value: primitive_value(
                        IDENT(
                            Token {
                                raw: "a",
                                position: Position {
                                    column: 5,
                                    line: 7,
                                },
                            },
                        ),
                    ),
                    expressions: [
                        expressions {
                            position: Position {
                                column: 7,
                                line: 7,
                            },
                            operator: PLUS(
                                Token {
                                    raw: "+",
                                    position: Position {
                                        column: 7,
                                        line: 7,
                                    },
                                },
                            ),
                            value: primitive_value(
                                IDENT(
                                    Token {
                                        raw: "b",
                                        position: Position {
                                            column: 9,
                                            line: 7,
                                        },
                                    },
                                ),
                            ),
                        },
                        expressions {
                            position: Position {
                                column: 11,
                                line: 7,
                            },
                            operator: PLUS(
                                Token {
                                    raw: "+",
                                    position: Position {
                                        column: 11,
                                        line: 7,
                                    },
                                },
                            ),
                            value: primitive_value(
                                float(
                                    float {
                                        position: Position {
                                            column: 13,
                                            line: 7,
                                        },
                                        whole: NUMBER(
                                            Token {
                                                raw: "0",
                                                position: Position {
                                                    column: 13,
                                                    line: 7,
                                                },
                                            },
                                        ),
                                        float: NUMBER(
                                            Token {
                                                raw: "012",
                                                position: Position {
                                                    column: 15,
                                                    line: 7,
                                                },
                                            },
                                        ),
                                    },
                                ),
                            ),
                        },
                    ],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 9,
        },
        statement: variable(
            variable {
                position: Position {
                    column: 0,
                    line: 9,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 9,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 9,
                                },
                            },
                        ),
                    },
                    modifiers {
                        position: Position {
                            column: 4,
                            line: 9,
                        },
                        modifier: MUT(
                            Token {
                                raw: "mut",
                                position: Position {
                                    column: 4,
                                    line: 9,
                                },
                            },
                        ),
                    },
                ],
                name: IDENT(
                    Token {
                        raw: "Pi",
                        position: Position {
                            column: 8,
                            line: 9,
                        },
                    },
                ),
                body: body {
                    position: Position {
                        column: 13,
                        line: 9,
                    },
                    value: primitive_value(
                        NUMBER(
                            Token {
                                raw: "123",
                                position: Position {
                                    column: 13,
                                    line: 9,
                                },
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 11,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 11,
                },
                modifiers: [],
                name: IDENT(
                    Token {
                        raw: "fn",
                        position: Position {
                            column: 0,
                            line: 11,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 3,
                            line: 11,
                        },
                        arguments: [
                            argument {
                                position: Position {
                                    column: 3,
                                    line: 11,
                                },
                                arg: IDENT(
                                    Token {
                                        raw: "a",
                                        position: Position {
                                            column: 3,
                                            line: 11,
                                        },
                                    },
                                ),
                            },
                        ],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 6,
                                    line: 11,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 11,
                        line: 11,
                    },
                    value: primitive_value(
                        IDENT(
                            Token {
                                raw: "a",
                                position: Position {
                                    column: 11,
                                    line: 11,
                                },
                            },
                        ),
                    ),
                    expressions: [
                        expressions {
                            position: Position {
                                column: 13,
                                line: 11,
                            },
                            operator: PLUS(
                                Token {
                                    raw: "+",
                                    position: Position {
                                        column: 13,
                                        line: 11,
                                    },
                                },
                            ),
                            value: primitive_value(
                                IDENT(
                                    Token {
                                        raw: "b",
                                        position: Position {
                                            column: 15,
                                            line: 11,
                                        },
                                    },
                                ),
                            ),
                        },
                    ],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 13,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 13,
                },
                modifiers: [],
                name: IDENT(
                    Token {
                        raw: "one",
                        position: Position {
                            column: 0,
                            line: 13,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 4,
                            line: 13,
                        },
                        arguments: [],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 4,
                                    line: 13,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 9,
                        line: 13,
                    },
                    value: function_call(
                        function_call {
                            position: Position {
                                column: 9,
                                line: 13,
                            },
                            path: [
                                path {
                                    position: Position {
                                        column: 9,
                                        line: 13,
                                    },
                                    path: IDENT(
                                        Token {
                                            raw: "console",
                                            position: Position {
                                                column: 9,
                                                line: 13,
                                            },
                                        },
                                    ),
                                },
                            ],
                            name: IDENT(
                                Token {
                                    raw: "log",
                                    position: Position {
                                        column: 17,
                                        line: 13,
                                    },
                                },
                            ),
                            arguments: [
                                calls {
                                    position: Position {
                                        column: 21,
                                        line: 13,
                                    },
                                    argument: body {
                                        position: Position {
                                            column: 21,
                                            line: 13,
                                        },
                                        value: primitive_value(
                                            IDENT(
                                                Token {
                                                    raw: "b",
                                                    position: Position {
                                                        column: 21,
                                                        line: 13,
                                                    },
                                                },
                                            ),
                                        ),
                                        expressions: [],
                                    },
                                },
                            ],
                        },
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 15,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 15,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 15,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 15,
                                },
                            },
                        ),
                    },
                ],
                name: IDENT(
                    Token {
                        raw: "zero",
                        position: Position {
                            column: 4,
                            line: 15,
                        },
                    },
                ),
                arguments: None,
                body: body {
                    position: Position {
                        column: 13,
                        line: 15,
                    },
                    value: primitive_value(
                        IDENT(
                            Token {
                                raw: "a",
                                position: Position {
                                    column: 13,
                                    line: 15,
                                },
                            },
                        ),
                    ),
                    expressions: [
                        expressions {
                            position: Position {
                                column: 15,
                                line: 15,
                            },
                            operator: PLUS(
                                Token {
                                    raw: "+",
                                    position: Position {
                                        column: 15,
                                        line: 15,
                                    },
                                },
                            ),
                            value: primitive_value(
                                IDENT(
                                    Token {
                                        raw: "b",
                                        position: Position {
                                            column: 17,
                                            line: 15,
                                        },
                                    },
                                ),
                            ),
                        },
                    ],
                },
            },
        ),
    },
]
